substratevm/src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/phases/NativeImageInlineDuringParsingPlugin.java

```diff
@@ -0,0 +1,313 @@
+            sb.append(site).append(" -> ").append(callee.format("%h.%n(%p)"));
+            String newIndentation = indentation + "  ";
+            for (InvocationResultInline child : children.values()) {
+                sb.append(System.lineSeparator()).append(newIndentation);
+                child.append(sb, newIndentation);
+            }
+            return sb;
+        }
+    }
+
+    public static class InvocationData {
+        private final ConcurrentMap<AnalysisMethod, ConcurrentMap<Integer, InvocationResult>> data = new ConcurrentHashMap<>();
+
+        private ConcurrentMap<Integer, InvocationResult> bciMap(ResolvedJavaMethod method) {
+            AnalysisMethod key;
+            if (method instanceof AnalysisMethod) {
+                key = (AnalysisMethod) method;
+            } else {
+                key = ((HostedMethod) method).getWrapped();
+            }
+
+            return data.computeIfAbsent(key, unused -> new ConcurrentHashMap<>());
+        }
+
+        InvocationResult get(ResolvedJavaMethod method, int bci) {
+            return bciMap(method).get(bci);
+        }
+
+        InvocationResult putIfAbsent(ResolvedJavaMethod method, int bci, InvocationResult value) {
+            return bciMap(method).putIfAbsent(bci, value);
+        }
+
+        public void onCreateInvoke(GraphBuilderContext b, int invokeBci, boolean analysis, ResolvedJavaMethod callee) {
+            if (b.getDepth() == 0) {
+
+                if (callee != null && callee.equals(b.getMetaAccess().lookupJavaMethod(SubstrateClassInitializationPlugin.ENSURE_INITIALIZED_METHOD))) {
+                    return;
+                }
+
+                ConcurrentMap<Integer, InvocationResult> map = bciMap(b.getMethod());
+                if (analysis) {
+                    map.putIfAbsent(invokeBci, InvocationResult.NO_ANALYSIS);
+                } else {
+                    InvocationResult state = map.get(invokeBci);
+                    if (state != InvocationResult.ANALYSIS_TOO_COMPLICATED && state != InvocationResult.NO_ANALYSIS) {
+                        throw VMError.shouldNotReachHere("Missing information for call site: " + b.getMethod().asStackTraceElement(invokeBci));
+                    }
+                }
+            }
+        }
+    }
+
+
+    private final boolean analysis;
+    private final HostedProviders providers;
+
+    public NativeImageInlineDuringParsingPlugin(boolean analysis, HostedProviders providers) {
+        this.analysis = analysis;
+        this.providers = providers;
+    }
+
+    @Override
+    public InlineInfo shouldInlineInvoke(GraphBuilderContext b, ResolvedJavaMethod method, ValueNode[] args) {
+       /* InvocationData data = ((SharedBytecodeParser) b).inlineInvocationData;
+        if (data == null) {
+            throw VMError.shouldNotReachHere("must not use SubstrateInlineDuringParsingPlugin when bytecode parser does not have InvocationData");
```

Comment:
it's a problem: throws VMError while parsing some methods, for example com.oracle.svm.core.graal.jdk.ArraycopySnippets.doArraycopy

Reply:
`doArraycopy` is a snippet and it should never end up being inlined in the first place. We have to filter it earlier.